### PR TITLE
Deal with starting / stopping weave

### LINKED ISSUE #510: Make scope deal with weave restarts (or weave starts after scope)
```
2015/09/23 06:07:17 error doing ticker: Get http://172.17.0.1:6784/report: dial tcp 172.17.0.1:6784: no route to host
2015/09/23 06:07:18 report generation took too long (4.194702594s)
2015/09/23 06:07:20 error doing ticker: Get http://172.17.0.1:6784/report: dial tcp 172.17.0.1:6784: no route to host
2015/09/23 06:07:21 report generation took too long (3.137524219s)
2015/09/23 06:07:24 error doing ticker: Get http://172.17.0.1:6784/report: dial tcp 172.17.0.1:6784: no route to host
2015/09/23 06:07:25 report generation took too long (4.42816017s)
2015/09/23 06:07:27 error doing ticker: Get http://172.17.0.1:6784/report: dial tcp 172.17.0.1:6784: no route to host
2015/09/23 06:07:28 report generation took too long (2.781171793s)
2015/09/23 06:07:31 error doing ticker: Get http://172.17.0.1:6784/report: dial tcp 172.17.0.1:6784: no route to host
2015/09/23 06:07:32 report generation took too long (4.325375019s)
2015/09/23 06:07:34 error doing ticker: Get http://172.17.0.1:6784/report: dial tcp 172.17.0.1:6784: no route to host
2015/09/23 06:07:36 report generation took too long (3.229815901s)
```

### LINKED ISSUE #784: TestWeaveTaggerOverlayTopology is flaky
@davkal saw this, now I've just seen it.

```
--- FAIL: TestWeaveTaggerOverlayTopology (0.00s)
	weave_test.go:69: 
		--- want
		+++ have
		@@ -14,20 +14,13 @@
		  Container: (report.Topology) {
		   Nodes: (report.Nodes) (len=1) {
		    (string) (len=18) "host1;83183a667c01": (report.Node) {
		-    Metadata: (report.Metadata) (len=3) {
		+    Metadata: (report.Metadata) (len=2) {
		      (string) (len=19) "docker_container_id": (string) (len=12) "83183a667c01",
		-     (string) (len=18) "weave_dns_hostname": (string) (len=20) "hostname.weave.local",
		-     (string) (len=17) "weave_mac_address": (string) (len=17) "d6:f2:5a:12:36:a8"
		+     (string) (len=18) "weave_dns_hostname": (string) (len=20) "hostname.weave.local"
		     },
		     Counters: (report.Counters) {
		     },
		-    Sets: (report.Sets) (len=2) {
		-     (string) (len=20) "docker_container_ips": (report.StringSet) (len=1 cap=1) {
		-      (string) (len=10) "10.0.0.123"
		-     },
		-     (string) (len=32) "docker_container_ips_with_scopes": (report.StringSet) (len=1 cap=1) {
		-      (string) (len=11) ";10.0.0.123"
		-     }
		+    Sets: (report.Sets) {
		     },
		     Adjacency: (report.IDList) <nil>,
		     Edges: (report.EdgeMetadatas) {
		
FAIL
FAIL	github.com/weaveworks/scope/probe/overlay	0.066s
```

### LINKED ISSUE #825: Controls websocket disconnects and doesn't reconnect
This happens in master after simply running `scope launch` (app and probe run in the same container):

```
<probe> 2016/01/15 15:44:25 probe starting, version 1133daa, ID 4ae281cc9b64c247
<probe> 2016/01/15 15:44:25 publishing to: localhost:4040, scope.weave.local:4040
<probe> 2016/01/15 15:44:25 Error fetching app details: Get http://127.0.0.1:4040/api: dial tcp 127.0.0.1:4040: getsockopt: connection refused
<probe> 2016/01/15 15:44:25 Control connection to 127.0.0.1:4040 starting
<probe> 2016/01/15 15:44:25 Error doing controls for 127.0.0.1:4040, backing off 1s: dial tcp 127.0.0.1:4040: getsockopt: connection refused
<app> 2016/01/15 15:44:25 app starting, version 1133daa, ID 7e93459ace052e6d
<app> 2016/01/15 15:44:25 listening on :4040
<probe> 2016/01/15 15:44:25 docker container: collecting stats for 25a1ca2a2001ef9e4c235c1af7dbb4ff4f718df257ffa63a9bebb93ba8ef1ffb
<probe> 2016/01/15 15:44:25 docker container: collecting stats for 895d1528f6681d30eafee0cfa1acd3a20bc590a05542a9140fdb5b5614827aa0
<probe> 2016/01/15 15:44:25 Publish loop for 127.0.0.1:4040 starting
<probe> 2016/01/15 15:44:25 docker container: collecting stats for 6848865c4c957666ef91d9cbf7a3f173133d354ece884ffbd5971aca36f6a800
<probe> 2016/01/15 15:44:25 docker container: collecting stats for 20b4673268e562bb5c87c028eacd27be10a61998eeea41408a1aeff74e2c37b6
<probe> 2016/01/15 15:44:25 docker container: collecting stats for cb68fa7cb9e5825d71bbc147b77bf88c3fa61c38e5a9887eb7eea225d6847869
<probe> 2016/01/15 15:44:25 docker container: collecting stats for fb1158e67bac5757936743aad74c0504f846187f9396739fa0e7a4d500dca82f
<probe> 2016/01/15 15:44:25 docker container: collecting stats for 54a97f58e4eca814dee85040a3746ae685bd216ed3e14ae94adfb6078318435f
<probe> 2016/01/15 15:44:25 docker container: stopped collecting stats for 25a1ca2a2001ef9e4c235c1af7dbb4ff4f718df257ffa63a9bebb93ba8ef1ffb
<probe> 2016/01/15 15:44:25 docker container: collecting stats for d373d55ac326ffb2a3a78d3d1e0a488c38a7aac4c5c5f66d6273ecc35be4e397
<probe> 2016/01/15 15:44:25 docker container: collecting stats for fc1be778df642e5a8aa93914788273bcd77b53051caf0c331a141a14d2902f91
<probe> 2016/01/15 15:44:25 docker container: collecting stats for d3430dfc08aff467b409b39914bbbad88c5a929db73fbe1e90ced186f755f0a3
<probe> 2016/01/15 15:44:25 docker container: collecting stats for faf115d00c15c5162682dbacead987a279ccf5436b73ed732bea103c9fbaad31
<probe> 2016/01/15 15:44:25 docker container: collecting stats for a9733b121ed840942db4eb77a59c10c5809de1b3ecbda44f5f9a8aaf302e857c
<probe> 2016/01/15 15:44:25 docker container: collecting stats for e112dabbe44bac5853e1738f6b20ffd38a96c994289ab3bfe93071ad7cf5f2d7
<probe> 2016/01/15 15:44:25 docker container: collecting stats for bb77d82c0cfa5229debe6a841795823fd3ee5652db3374e53c900b416671100d
<probe> 2016/01/15 15:44:26 docker container: collecting stats for 61d2e2f57a4668b884db7cfbe64b66a18a96af775c55643eb2421f280491c627
<probe> 2016/01/15 15:44:26 docker container: collecting stats for 1e54d281f3555a061999be9577117eb961b88052292acd002e70675229f972da
<probe> 2016/01/15 15:44:26 docker container: collecting stats for ee6e86a8d69640dff7cff12177e44af55f74f4ef1cad61d5d60fa65100fe5cf6
<probe> 2016/01/15 15:44:26 docker container: collecting stats for 8e17303c5a95b93315b319adcf05d88446203f6d21e0948be17fef63a8108a6e
<probe> 2016/01/15 15:44:26 docker container: collecting stats for 7513b72be0f04f367078ff35087a5fe0be89b495549b13f00e665f108ac92106
<probe> 2016/01/15 15:44:26 docker container: collecting stats for b0d2d512d6b2816d7c7553609a5c08c3fa7de919e0b878074c0f2f54f513d50c
<probe> 2016/01/15 15:44:35 Closing control connection to 127.0.0.1:4040
<probe> 2016/01/15 15:44:35 Control connection to 127.0.0.1:4040 starting
<probe> 2016/01/15 15:44:35 Publish loop for 127.0.0.1:4040 exiting
<probe> 2016/01/15 15:44:35 Closing control connection to 127.0.0.1:4040
<probe> 2016/01/15 15:44:35 Control connection to 127.0.0.1:4040 exiting
<probe> 2016/01/15 15:44:37 Publish loop for 127.0.0.1:4040 starting
<app> 2016/01/15 16:55:20 Probe 4ae281cc9b64c247 is not connected right now...
<app> 2016/01/15 16:55:21 Probe 4ae281cc9b64c247 is not connected right now...
<app> 2016/01/15 16:55:22 Probe 4ae281cc9b64c247 is not connected right now...
<app> 2016/01/15 16:55:23 Probe 4ae281cc9b64c247 is not connected right now...
<app> 2016/01/15 16:55:23 Probe 4ae281cc9b64c247 is not connected right now...
<app> 2016/01/15 16:55:23 Probe 4ae281cc9b64c247 is not connected right now...
<app> 2016/01/15 16:55:24 Probe 4ae281cc9b64c247 is not connected right now...
<app> 2016/01/15 16:55:24 Probe 4ae281cc9b64c247 is not connected right now...
<app> 2016/01/15 16:55:25 Probe 4ae281cc9b64c247 is not connected right now...
<app> 2016/01/15 16:55:25 Probe 4ae281cc9b64c247 is not connected right now...
<app> 2016/01/15 16:55:26 Probe 4ae281cc9b64c247 is not connected right now...
```

The `Probe 4ae281cc9b64c247 is not connected right now` happens whenever I try to use a control from the UI.

Note how the probe disconnects (`Control connection to 127.0.0.1:4040 exiting`) and never connects again.

### PR DESCRIPTION
**Pull request by tomwilkie**
Fixes #510, fixes #784, fixes #825

A few things rolled up in this:
- Factor out all the stuff that talks to weave into a client library
- Have everything try and talk to weave by default (ie its no longer enabled iff weave is detected at startup)
- Have everything that talks to weave back off if it fails; when it hits max back off, we stop logging about it.  All the back off code is common.
- Have the app register with weave dns, and the probe poll system dns and weave dns 

This has a few nice effects:
- You can start weave after scope, at it should still cluster up automatically. 
- Similarly you can restart weave with no problems.
- You can now split the app and the probe on a weave network, at it will all behave as expected (ie `weave scope launch --no-app` on every machine and `weave scope launch --no-probe` on a few)

### CODE REVIEW

**Comment by tomwilkie** [comment]
@paulbellamy I'm working through the test breakages, but I think this is ready to be reviewed.  I'd really like to get this in before I go off on holiday (Friday)

File: prog/probe.go [hunk 1]
```diff
@@ -169,3 +184,22 @@ func probeMain() {
...
+	if err != nil {
+		return "", err
+	}
+
+	addrs, err := inf.Addrs()
+	if err != nil {
+		return "", err
+	}
+
+	switch v := addrs[0].(type) {
```
Comment (paulbellamy) [inline prog/probe.go]: addrs will never be empty?

**Comment by paulbellamy** [comment]
This doesn't seem to be attributing weave endpoints to containers. It seems to just attribute them to "TheInternet".

**Comment by tomwilkie** [comment]
Sweet, you found the bug!

**Comment by tomwilkie** [comment]
Should also use the weave client api: https://github.com/weaveworks/weave/tree/master/api

**Comment by tomwilkie** [comment]
I can't explain why the coverage has dropped so much, all the new code has reasonably good coverage.

**Pull request merged**